connector/ws/src/lib.rs

```diff
@@ -363,16 +371,27 @@ impl<S: connector::Supervisor> Inner<S> {
         let mut last_received_ping = tokio::time::Instant::now();
 
         loop {
-            // Check if we should exit** because (1) a shutdown is requested
-            // and (2) we are currently idle (no job is running).
-            if self.shutdown_requested.load(Ordering::SeqCst) {
-                let st = self.last_updated_status.lock().await.clone();
-                if matches!(st, ReportedSupervisorStatus::Idle) {
-                    tracing::info!("Shutdown requested, and no job is active; exiting run().");
-                    return;
-                }
+            // Check if the watch channel has changed
+            if *self.shutdown_rx.borrow() && self.is_idle().await {
+                tracing::info!("Shutdown requested, and supervisor is idle; exiting run().");
+                return;
             }
+
             tokio::select! {
+                // Check if the watch channel has changed
+                changed_result = shutdown_rx.changed() => {
+                    if changed_result.is_ok() {
+                        // We got a new value in the watch channel. Check if it's `true`:
+                        let shutdown_requested = *self.shutdown_rx.borrow();
+                        if shutdown_requested && self.is_idle().await {
+                            tracing::info!("Shutdown requested (watch changed) and idle => exit run().");
+                            return;
+                        }
+                    } else {
+                        // If `changed()` errors, it means all watch senders have dropped. Probably safe to ignore.
+                        tracing::warn!("shutdown_rx.changed() returned an error (sender dropped?)");
+                    }
+                }
```

Comment:
I think this all looks a bit redundant with what Tokio provides in the `wait_for` method: https://docs.rs/tokio/latest/tokio/sync/watch/struct.Receiver.html#method.wait_for

We should be able to just convert this select branch into a `wait_for` query and then even get rid of the if-block above the select branch, no?